src/storage/secondary/RedisStorage.cpp

```diff
@@ -0,0 +1,353 @@
+// this program; if not, write to the Free Software Foundation, Inc., 51
+// Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+
+#include "RedisStorage.hpp"
+
+#include <AtomicFile.hpp>
+#include <Digest.hpp>
+#include <Logging.hpp>
+#include <UmaskScope.hpp>
+#include <Util.hpp>
+#include <assertions.hpp>
+#include <fmtmacros.hpp>
+#include <util/file_utils.hpp>
+#include <util/string_utils.hpp>
+
+#include <third_party/nonstd/string_view.hpp>
+
+namespace storage {
+namespace secondary {
+
+static struct timeval
+milliseconds_to_timeval(const std::string& msec)
+{
+  int ms = std::stoi(msec);
+  struct timeval tv;
+  tv.tv_sec = ms / 1000;
+  tv.tv_usec = (ms % 1000) * 1000;
+  return tv;
+}
+
+static std::string
+timeval_to_string(struct timeval tv)
+{
+  return FMT("{:.3f}s", tv.tv_sec + tv.tv_usec / 1000000.0);
+}
+
+static nonstd::optional<struct timeval>
+parse_connect_timeout(const AttributeMap& attributes)
+{
+  const auto it = attributes.find("connect-timeout");
+  if (it == attributes.end()) {
+    return nonstd::nullopt;
+  }
+  return milliseconds_to_timeval(it->second);
+}
+
+static nonstd::optional<struct timeval>
+parse_operation_timeout(const AttributeMap& attributes)
+{
+  const auto it = attributes.find("operation-timeout");
+  if (it == attributes.end()) {
+    return nonstd::nullopt;
+  }
+  return milliseconds_to_timeval(it->second);
+}
+
+static nonstd::optional<std::string>
+parse_username(const AttributeMap& attributes)
+{
+  const auto it = attributes.find("username");
+  if (it == attributes.end()) {
+    return nonstd::nullopt;
+  }
+  return it->second;
+}
+
+static nonstd::optional<std::string>
+parse_password(const AttributeMap& attributes)
+{
+  const auto it = attributes.find("password");
+  if (it == attributes.end()) {
+    return nonstd::nullopt;
+  }
+  return it->second;
+}
+
+RedisStorage::RedisStorage(const std::string& url,
+                           const AttributeMap& attributes)
+  : m_url(url),
+    m_connect_timeout(parse_connect_timeout(attributes)),
+    m_operation_timeout(parse_operation_timeout(attributes)),
+    m_username(parse_username(attributes)),
+    m_password(parse_password(attributes))
+{
+  m_context = nullptr;
+  m_connected = false;
+  m_invalid = false;
+}
+
+RedisStorage::~RedisStorage()
+{
+  disconnect();
+  if (m_context) {
+    redisFree(m_context);
+    m_context = nullptr;
+  }
+}
+
+bool
+RedisStorage::connect()
+{
+  if (m_connected) {
+    return true;
+  }
+  if (m_invalid) {
+    return false;
+  }
+
+  if (m_context) {
+    if (redisReconnect(m_context) == REDIS_OK) {
+      m_connected = true;
+      return true;
+    }
+    LOG("Redis reconnect err: {}", m_context->errstr);
+    redisFree(m_context);
+    m_context = nullptr;
+  }
+
+  const nonstd::string_view prefix = "redis://";
+  ASSERT(Util::starts_with(m_url, prefix));
+  const auto suffix = m_url.substr(prefix.size());
+  std::string host, port, sock;
```

Comment:
Hopefully we can use the `Url` class added in #676 for this instead.

Reply:
I have a preparational PR to add the Url class in #881